File: specs/json-schema/research/format_typed_repr/Typed.cs
// Probe: .NET (C#) STANDARD LIBRARY (BCL) typed reps for the 6 formats.
// BCL only (System.*). Run via: dotnet run  (see typed_cs.csproj)
// Backs features/format typed-repr research. .NET is a PROSPECTIVE target.
using System;
using System.Globalization;
using System.Net;
using System.Net.Sockets;

static void Line(string label, string s, Func<string> f) {
    try { Console.WriteLine($"  {label,-12} {("\"" + s + "\""),-42} -> {f()}"); }
    catch (Exception e) { Console.WriteLine($"  {label,-12} {("\"" + s + "\""),-42} -> ERR {e.GetType().Name}: {e.Message}"); }
}

Console.WriteLine("=== .NET BCL typed representations ===");

// ---- date-time : DateTimeOffset (preserves offset) ----
Console.WriteLine("\n[date-time] type=System.DateTimeOffset  ctor=DateTimeOffset.Parse(s, InvariantCulture, RoundtripKind)");
foreach (var s in new[]{
    "2021-02-28T23:59:60Z",            // leap second
    "2006-01-02T15:04:05Z",
    "2006-01-02T15:04:05+00:00",
    "2006-01-02T15:04:05-00:00",
    "2006-01-02T15:04:05.123456789Z",  // 9-digit
    "2006-01-02t15:04:05z",            // lowercase
    "2006-01-02T15:04:05",             // missing offset
    "2021-02-30T00:00:00Z",            // bad calendar
}) {
    Line("DateTimeOffset", s, () => {
        var d = DateTimeOffset.Parse(s, CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind);
        return $"OK -> {d:o}";
    });
}

// ---- date : DateOnly (net6+) ----
Console.WriteLine("\n[date] type=System.DateOnly (net6+)  ctor=DateOnly.ParseExact(s, \"yyyy-MM-dd\")");
foreach (var s in new[]{"2020-02-29", "2021-02-29", "2021-13-01"})
    Line("DateOnly", s, () => "OK -> " + DateOnly.ParseExact(s, "yyyy-MM-dd", CultureInfo.InvariantCulture).ToString("yyyy-MM-dd"));

// ---- time : TimeOnly (net6+) -- but RFC3339 time carries an offset TimeOnly cannot hold ----
Console.WriteLine("\n[time] type=System.TimeOnly (net6+). NOTE: TimeOnly has NO offset; RFC3339 time offset would be lost.");
foreach (var s in new[]{"12:00:00", "23:59:60", "12:00:00.5"})
    Line("TimeOnly", s, () => "OK -> " + TimeOnly.ParseExact(s, new[]{"HH:mm:ss","HH:mm:ss.fffffff","HH:mm:ss.f"}, CultureInfo.InvariantCulture).ToString());

// ---- uuid : System.Guid ----
Console.WriteLine("\n[uuid] type=System.Guid  ctor=Guid.Parse(s) / Guid.ParseExact(s, \"D\")");
foreach (var s in new[]{
    "f81d4fae-7dec-11d0-a765-00a0c91e6bf6",
    "F81D4FAE-7DEC-11D0-A765-00A0C91E6BF6",
    "f81d4fae7dec11d0a76500a0c91e6bf6",      // no dashes (Parse accepts! ParseExact D rejects)
    "{f81d4fae-7dec-11d0-a765-00a0c91e6bf6}",// braces (Parse accepts!)
    "not-a-guid"}) {
    Line("Guid.Parse", s, () => "OK -> " + Guid.Parse(s).ToString());
}
Console.WriteLine("  -- Guid.ParseExact(s, \"D\") strict hyphenated form --");
foreach (var s in new[]{"f81d4fae-7dec-11d0-a765-00a0c91e6bf6", "F81D4FAE-7DEC-11D0-A765-00A0C91E6BF6",
                        "f81d4fae7dec11d0a76500a0c91e6bf6", "{f81d4fae-7dec-11d0-a765-00a0c91e6bf6}"})
    Line("ParseExact D", s, () => "OK -> " + Guid.ParseExact(s, "D").ToString());

// ---- ipv4 / ipv6 : System.Net.IPAddress ----
Console.WriteLine("\n[ipv4/ipv6] type=System.Net.IPAddress  ctor=IPAddress.Parse(s)");
foreach (var s in new[]{"192.168.0.1", "256.0.0.1", "01.2.3.4", "1.2.3",
                        "::1", "2001:db8::1", "2001:DB8::1", "::ffff:192.168.0.1", "fe80::1%12"}) {
    Line("IPAddress", s, () => {
        var a = IPAddress.Parse(s);
        return $"OK fam={a.AddressFamily} to_s={a}";
    });
}

File: specs/json-schema/research/format_typed_repr/typed.rb
# Probe: Ruby STANDARD LIBRARY typed reps for the 6 formats.
# Uses only stdlib requires (date, time, ipaddr, securerandom). Run: ruby typed.rb
# Backs features/format typed-repr research. Ruby is a PROSPECTIVE target.
require 'date'
require 'time'
require 'ipaddr'

def line(label, s)
  begin
    r = yield
    printf("  %-10s %-42s -> %s\n", label, s.inspect, r)
  rescue => e
    printf("  %-10s %-42s -> ERR %s: %s\n", label, s.inspect, e.class, e.message)
  end
end

puts "=== Ruby stdlib typed representations ==="

# ---- date-time : DateTime.rfc3339 / Time.iso8601 (require 'date' / 'time') ----
puts "\n[date-time] type=DateTime (or Time). ctor=DateTime.rfc3339(s) / Time.iso8601(s)"
[
  "2021-02-28T23:59:60Z",           # leap second
  "2006-01-02T15:04:05Z",
  "2006-01-02T15:04:05+00:00",
  "2006-01-02T15:04:05-00:00",
  "2006-01-02T15:04:05.123456789Z", # 9-digit
  "2006-01-02t15:04:05z",           # lowercase
  "2006-01-02T15:04:05",            # missing offset
  "2021-02-30T00:00:00Z",           # bad calendar
].each do |s|
  line("DateTime.rfc3339", s) { d = DateTime.rfc3339(s); "OK -> #{d.iso8601(9)}" }
end
puts "  -- Time.iso8601 --"
["2021-02-28T23:59:60Z", "2006-01-02T15:04:05", "2006-01-02T15:04:05+00:00"].each do |s|
  line("Time.iso8601", s) { t = Time.iso8601(s); "OK -> #{t.iso8601}" }
end

# ---- date : Date.iso8601 / Date.parse ----
puts "\n[date] type=Date  ctor=Date.iso8601(s) (strict) / Date.parse (lax)"
["2020-02-29", "2021-02-29", "2021-13-01"].each do |s|
  line("Date.iso8601", s) { "OK -> #{Date.iso8601(s).iso8601}" }
end

# ---- time : NO time-of-day-only stdlib type ----
puts "\n[time] NO time-of-day-only stdlib type. Time is a full instant; parsing '12:00:00' fills in today's date."
["12:00:00", "23:59:60Z"].each do |s|
  line("Time.parse", s) { Time.parse(s).iso8601 rescue (raise) }
end

# ---- uuid : NO stdlib UUID type. SecureRandom.uuid GENERATES a string only ----
puts "\n[uuid] NO stdlib UUID type. SecureRandom.uuid GENERATES a plain string; no parse/validate type."
require 'securerandom'
puts "  SecureRandom.uuid sample: #{SecureRandom.uuid} (a String, not a typed object)"

# ---- ipv4 / ipv6 : IPAddr (require 'ipaddr') ----
puts "\n[ipv4] type=IPAddr  ctor=IPAddr.new(s)"
["192.168.0.1", "256.0.0.1", "01.2.3.4", "1.2.3", "1.2.3.4.5"].each do |s|
  line("IPAddr", s) { a = IPAddr.new(s); "OK ipv4?=#{a.ipv4?} to_s=#{a.to_s}" }
end
puts "\n[ipv6] type=IPAddr  ctor=IPAddr.new(s)"
["::1", "2001:db8::1", "2001:DB8::1", "::ffff:192.168.0.1",
 "2001:0db8:0000:0000:0000:0000:0000:0001"].each do |s|
  line("IPAddr", s) { a = IPAddr.new(s); "OK ipv6?=#{a.ipv6?} to_s=#{a.to_s}" }
end
